File: common/brownshome/scriptwars/game/tanks/Coordinates.java
package brownshome.scriptwars.game.tanks;

import brownshome.scriptwars.client.Network;

public class Coordinates {
	private final int x;
	private final int y;
	
	public Coordinates(int x, int y){
		this.x = x;
		this.y = y;
	}
	
	public Coordinates(Network network) {
		this(network.getByte(), network.getByte()); //The execution order here is guaranteed
	}

	public int getX() {
		return x;
	}
	
	public int getY() {
		return y;
	}
	
	@Override
	public boolean equals(Object obj) {
		if(!(obj instanceof Coordinates)) {
			return false;
		}
		
		Coordinates other = (Coordinates) obj;

		return other.x == x && other.y == y;
	}

	@Override
	public int hashCode() {
		return x ^ (y * 31); //TODO distribute high order bits better
	}
	
	@Override
	public String toString() {
		return "(" + x + ", " + y + ")";
	}

	public static Coordinates getRandom(int maxX, int maxY) {
		return new Coordinates((int) (Math.random() * maxX), (int) (Math.random() * maxY));
	}
}

File: client/python/Coordinates.py
class Coordinates:
    def __init__(self, network=None):
        if network is not None:
            self.x = network.getByte()
            self.y = network.getByte()
        
    def __hash__(self):
        return self.x ^ self.y
        
    def __eq__(self, other):
        return (self.x, self.y) == (other.x, other.y)
        
    def __ne__(self, other):
        return not(self == other)
